icu4j/main/core/src/main/java/com/ibm/icu/segmenter/SegmentsImpl.java

```diff
@@ -0,0 +1,128 @@
+    // which then gets cloned and used in each of the Segments APIs' implementations. The second
+    // level of cloning that happens when the Segments object's local BreakIterator prototype
+    // gets cloned allows the iteration state to be separate whenever an Segments API is called.
+    // Otherwise, there is a chance that multiple API calls on the same Segments object might
+    // mutate the same position/index, if done concurrently.
+    breakIterPrototype = (BreakIterator) breakIter.clone();
+    // It's okay to perform .setText on the object that we want to clone later because we should
+    // then not have to call .setText on the clones.
+    breakIterPrototype.setText(source);
+  }
+
+  @Override
+  public Segment segmentAt(int i) {
+    BreakIterator breakIter = (BreakIterator) breakIterPrototype.clone();
+    int start;
+    int limit;
+
+    boolean isBoundary = breakIter.isBoundary(i);
+
+    if (isBoundary) {
+      start = i;
+      limit = breakIter.next();
+    } else {
+      // BreakIterator::isBoundary(i) will advance forwards to the next boundary if the argument
+      // is not a boundary.
+      limit = breakIter.current();
+      start = breakIter.previous();
+    }
+
+    if (start != BreakIterator.DONE && limit != BreakIterator.DONE) {
+      return new Segment(start, limit, source);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public Stream<Segment> segments() {
+    BreakIterator breakIter = (BreakIterator) breakIterPrototype.clone();
+    breakIter.setText(source);
+
+    // create a Stream from a Spliterator of an Iterable so that the Stream can be lazy, not eager
+    SegmentIterable iterable = new SegmentIterable(breakIter, IterationDirection.FORWARDS, 0,
+        source);
+    return StreamSupport.stream(iterable.spliterator(), false);
+  }
+
+  @Override
+  public boolean isBoundary(int i) {
+    return ((BreakIterator) breakIterPrototype.clone()).isBoundary(i);
+  }
+
+  @Override
+  public Stream<Segment> segmentsFrom(int i) {
+    BreakIterator breakIter = (BreakIterator) breakIterPrototype.clone();
+    breakIter.setText(source);
+
+    // create a Stream from a Spliterator of an Iterable so that the Stream can be lazy, not eager
+    SegmentIterable iterable = new SegmentIterable(breakIter, IterationDirection.FORWARDS, i,
+        source);
+    return StreamSupport.stream(iterable.spliterator(), false);
+  }
+
+  @Override
+  public Stream<Segment> segmentsBefore(int i) {
+    BreakIterator breakIter = (BreakIterator) breakIterPrototype.clone();
+    breakIter.setText(source);
+
+    // create a Stream from a Spliterator of an Iterable so that the Stream can be lazy, not eager
+    SegmentIterable iterable = new SegmentIterable(breakIter, IterationDirection.BACKWARDS, i,
+        source);
+    return StreamSupport.stream(iterable.spliterator(), false);
+  }
+
+  @Override
+  public IntStream boundaries() {
```

Comment:
call boundariesAfter(-1) to reduce code?

maybe even do so in the interface definition?